cf/weights.py

```diff
@@ -0,0 +1,1903 @@
+            # Spherical polygons defined by great circles
+            # --------------------------------------------------------
+            if not great_circle:
+                raise ValueError(
+                    "Must set great_circle=True to allow the derivation "
+                    "of volume weights from spherical polygons composed "
+                    "from great circle segments."
+                )
+
+            if methods:
+                weights[(axis,)] = "volume spherical polygon geometry"
+                return True
+
+            if measure:
+                r = f.radius(default=radius)
+
+                # actual_volume =
+                #    [actual_area/(4*pi*r**2)]
+                #    * (4/3)*pi*[(r+delta_z)**3 - r**3)]
+                volumes = areas * (
+                    delta_z**3 / (3 * r**2) + delta_z**2 / r + delta_z
+                )
+            else:
+                volumes = areas
+        else:
+            raise ValueError(
+                "X and Y coordinate bounds must both have units "
+                "equivalent to either metres (for plane polygon) or "
+                "radians (for spherical polygon) volume calculations. Got "
+                f"{x.Units!r} and {y.Units!r}."
+            )
+
+        weights[(axis,)] = volumes
+        weights_axes.add(axis)
+        return True
+
+    @classmethod
+    def interior_angles(cls, f, lon, lat, interior_rings=None):
+        """Find the interior angles at spherical polygon vertices.
+
+        The interior angle at a vertex is that formed by two adjacent
+        sides. Each interior angle is therefore a function of three
+        vertices - the vertex at which the interior angle is required and
+        the two adjacent vertices either side of it.
+
+        **Method**
+
+        Bevis, M., Cambareri, G. Computing the area of a spherical polygon
+          of arbitrary shape. Math Geol 19, 335–346 (1987).
+
+        http://bemlar.ism.ac.jp/zhuang/Refs/Refs/bevis1987mathgeol.pdf
+
+        Abstract: An algorithm for determining the area of a spherical
+          polygon of arbitrary shape is presented. The kernel of the
+          problem is to compute the interior angle at each vertex of the
+          spherical polygon; a well-known relationship between the area of
+          a spherical polygon and the sum of its interior angles then may
+          be exploited. The algorithm has been implemented as a FORTRAN
+          subroutine and a listing is provided.
+
+        .. versionadded:: 3.16.0
+
+        .. seealso:: `central_angles`
+
+        :Parameters:
+
+            f: `Field`
+                The field for which the weights are being created.
+
+            lon: `Data`
+                Longitudes. Must have units of radians, which is not
+                checked.
+
+            lat: `Data`
+                Latitudes. Must have units of radians, which is not
+                checked.
```

Comment:
Any reason we don't check units here, since it is of course simple enough to do and we verify elsewhere in the code very often?

Reply:
If we're calling this, then we will already have checked this, so no need to do so again. It's similar for `central_angles`. Problem solved by making them private methods: 09d9ba5363e53a329c7fec0080c7dc5caea31eee

Reply:
Makes sense. Thanks for clarifying.